### PR TITLE
Execute task assert program address

### PR DESCRIPTION
**Pull request by notlesh**
Execute task hint 4

I'm assuming the type of `program_address` is a `Relocatable`, let me know if that's wrong...

## Checklist
- [ ] Linked to Github Issue
- [x] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
  - [ ] CHANGELOG has been updated.

### CODE REVIEW

File: vm/src/hint_processor/builtin_hint_processor/bootloader/bootloader_hints.rs [hunk 1]
```diff
@@ -332,6 +332,33 @@ pub fn guess_pre_image_of_subtasks_output_hash(
...
+    assert ids.program_address == program_address"
+%}
+*/
+pub fn assert_program_address(
+    vm: &mut VirtualMachine,
+    exec_scopes: &mut ExecutionScopes,
+    ids_data: &HashMap<String, HintReference>,
+    ap_tracking: &ApTracking,
+) -> Result<(), HintError> {
+    let ids_program_address = get_ptr_from_var_name("program_address", vm, ids_data, ap_tracking)?;
```
Comment (notlesh) [inline vm/src/hint_processor/builtin_hint_processor/bootloader/bootloader_hints.rs:348]: TODO: use `mod vars` convention

File: vm/src/hint_processor/builtin_hint_processor/bootloader/bootloader_hints.rs [hunk 2]
```diff
@@ -332,6 +332,33 @@ pub fn guess_pre_image_of_subtasks_output_hash(
...
+    let program_address = exec_scopes.get_any_boxed_ref("program_address")?;
+    let program_address = program_address
+        .downcast_ref::<Relocatable>()
+        .ok_or(HintError::WrongHintData)?;
```
Comment (odesenfans) [inline vm/src/hint_processor/builtin_hint_processor/bootloader/bootloader_hints.rs:352]:
```suggestion
    let program_address: Relocatable = exec_scopes.get("program_address")?;
```

File: vm/src/hint_processor/builtin_hint_processor/bootloader/bootloader_hints.rs [hunk 3]
```diff
@@ -332,6 +332,33 @@ pub fn guess_pre_image_of_subtasks_output_hash(
...
+
+    if &ids_program_address != program_address {
```
Comment (odesenfans) [inline vm/src/hint_processor/builtin_hint_processor/bootloader/bootloader_hints.rs:354]: Why are you using a reference? Both are `Relocatable` objects.

File: vm/src/hint_processor/builtin_hint_processor/bootloader/bootloader_hints.rs [hunk 4]
```diff
@@ -832,4 +859,58 @@ mod tests {
...
+
+        let mut new_segment_base = vm.add_memory_segment();
+        let _ = insert_value_from_var_name(
+            "program_address",
+            new_segment_base.clone(),
+            &mut vm,
+            &ids_data,
+            &ap_tracking,
+        )
+        .map_err(|e| panic!("could not insert var: {}", e));
```
Comment (odesenfans) [inline vm/src/hint_processor/builtin_hint_processor/bootloader/bootloader_hints.rs:884]: You don't really need to add a new segment. You can just insert an arbitrary relocatable value in `ids.program_address` and `exec_scopes` without having an actual address backing it up (i.e. the pointer does not need to be valid, nor point to a new segment). If you want a valid pointer, you can also just point to some address in the already created segments, like (0, 0) or (1, 0).

File: vm/src/hint_processor/builtin_hint_processor/bootloader/bootloader_hints.rs [hunk 5]
```diff
@@ -832,4 +859,58 @@ mod tests {
...
+            ids_data,
+        );
+        let mut hint_processor = BuiltinHintProcessor::new_empty();
+
+        let result = hint_processor.execute_hint(
+            &mut vm,
+            &mut exec_scopes,
+            &any_box!(hint_data),
+            &HashMap::new(),
+        );
```
Comment (odesenfans) [inline vm/src/hint_processor/builtin_hint_processor/bootloader/bootloader_hints.rs:904]: While I'm not strictly against doing it this way, using `run_hint!` feels more straight to the point.

**Review by odesenfans — CHANGES_REQUESTED** [review]

File: vm/src/hint_processor/builtin_hint_processor/bootloader/bootloader_hints.rs (hunk 4 above)
Comment (notlesh) [reply vm/src/hint_processor/builtin_hint_processor/bootloader/bootloader_hints.rs]: Good point. Fixed in dae546b

File: vm/src/hint_processor/builtin_hint_processor/bootloader/bootloader_hints.rs (hunk 5 above)
Comment (notlesh) [reply vm/src/hint_processor/builtin_hint_processor/bootloader/bootloader_hints.rs]: Good point, it's better here.

**Pull request merged**